### ISSUE TITLE
Sending to clients failing with multiple clients after an extended period of time. 

### ISSUE THREAD

**Issue by Bernstern** [miguelgrinberg/Flask-SocketIO#1675]
**Describe the bug**
We have begun to notice a bug with our backend setup where after an extended period of time, sending data to clients will no longer occur. Calling `send(...)` itself raises no errors but we stop seeing a log that the underlying emit is called. 

At least we don't think that the emit is actually occuring because the log line below in the pubsub manager never executes:
```
self._get_logger().info('pubsub message: {}'.format(...)
```
And clients who are still receiving ping-pong messages from the backend never receive the emitted message. 

**To Reproduce**
1. We have not found a consistent way to trigger the bug other as it seems to happen after a week or so between restarts but we will include that if we find a way to trigger it. 

**Expected behavior**
Emits to work with multiple clients maintaining a web socket connection. 

**Logs**
When it is working we see logs like so:
```
socketio.server INFO    emitting event "message" to <room name> [/<namespace>]
socketio.server INFO    pubsub message: emit
```
Whenever the bug occurs we no longer see the pubsub message. 

**Additional context**
We are using flask-socketio on aws with elastic beanstalk (with an application load balancer) with redis for the message queue and gevent as the async mode setup below:
```
socketio = SocketIO(
    app,
    async_mode="gevent",
    cors_allowed_origins="*",
    manage_session=False,
    message_queue="redis://...",
    logger=True,
    engine_logger=True,
)
```
We are using the following package versions:
```
python-socketio==5.3.0
gevent-websocket==0.10.1
Flask-SocketIO==5.1.0
redis==3.5.3
gevent==21.1.1
```

We have also found that restarting the elastic beanstalk environment fixes the issue temporarily but we are not sure why. We originally thought it was an issue with amazon replacing ec2 instances in our EB, but even with that disabled we still see this behavior. We also verified that clients are using websockets instead of long polling.

**Comment by miguelgrinberg** [comment]
I need to see complete logs covering a period of time including the last message sent that worked, and the first that didn't. Is that possible?

**Comment by Bernstern** [comment]
There's a lot of PII information in our backend logs so I don't believe I can include a dump of them, however we did track down the last instance where an emit occurred successfully and then what happened in the next time the same interaction happened. 

Below is what we see when a client connects to our server and we emit two messages, one of which has a callback: 
```
socketio.server INFO    received event "join" from d0ccqVG-TqBIzb6wAAI3 [/python]
backend.resources.socket        INFO    Request to join resource room from 9676a712-ca4e-413a-be53-35ed2c45aac5
socketio.server INFO    d0ccqVG-TqBIzb6wAAI3 is entering room 9676a712-ca4e-413a-be53-35ed2c45aac5 [/python]
socketio.server INFO    emitting event "message" to 9676a712-ca4e-413a-be53-35ed2c45aac5 [/python]
socketio.server INFO    pubsub message: emit
socketio.server INFO    emitting event "message" to 9676a712-ca4e-413a-be53-35ed2c45aac5 [/python]
socketio.server INFO    pubsub message: emit
socketio.server INFO    received ack from d0ccqVG-TqBIzb6wAAI3 [/python]
socketio.server INFO    pubsub message: callback
```

After that and before restarting our backend we are left with logs that look like this where emits no longer occur: 
```
socketio.server INFO    received event "join" from fiUxwouCUjPvdhAMAAV2 [/python]
backend.resources.socket        INFO    Request to join resource room from 132f95b7-1bb2-4a54-8340-468e375c480e
socketio.server INFO    fiUxwouCUjPvdhAMAAV2 is entering room 132f95b7-1bb2-4a54-8340-468e375c480e [/python]
socketio.server INFO    emitting event "message" to 132f95b7-1bb2-4a54-8340-468e375c480e [/python]
socketio.server INFO    emitting event "message" to 132f95b7-1bb2-4a54-8340-468e375c480e [/python]
```

We are looking into checking our redis queue which is handled by elasticache right now but we did notice the following line getting logged both when we restart the server as well as roughly around when the sending stops:
```
socketio.server INFO    redis backend initialized.
```

We read through the logs after the last successful message above and there were no enginio or socketio.server logs.

**Comment by miguelgrinberg** [comment]
The `redis backend initialized` message appears when the clilent manager instance is created. I haven't seen your code, but I assume you are not creating your own `RedisClientManager` object? In that case, it's Flask-SocketIO creating it, and I don't see any situation in which a new client manager instance would be created during an active server session.

Without seeing your logs I can't really tell you anything more. Have you looked for Python stack traces? Or any redis retry messages from the python-socketio package?

**Comment by Bernstern** [comment]
That is correct, we are not directly initializing the `RwedisClientManager`, I can put together some more logs and get the relevant pieces of the source code together to add to this issue over the next few days. 

In the meantime we did notice that our Redis running on AWS elasticache was set with `Engine Version Compatibility = 3.2.10` which seems very old. I updated it to '6.x' to see if that helps with the issue. 

Looking through the logs there are no other redis logs, we don't see the trying to re-connect attempts and there are no stack traces/random exceptions from socketio/engineio.

**Comment by miguelgrinberg** [comment]
@Bernstern Based on your description of the problem the only logical explanation I can come up with is that a call to redis blocks and never returns. The `socket_timeout` argument of the redis connection defaults to `None`, which is then passed to the socket object to mean no timeout, so this supports the theory that for some reason your Redis at some point hangs. I have no idea what could make Redis get into this state though.

If you feel like experimenting, you can add a `socket_timeout` on the Redis connections created in the redis manager class. That may prevent the blocking state, but I'm not sure a new connection will work any better if the service is in a bad state.

**Comment by miguelgrinberg** [comment]
@llunn Are you logging stderr? I think what you are describing could be caused by an unhandled exception in the Redis listening thread, which now I see that does not have a catch all error handler, only for redis specific errors. But if this was your case, then you should have seen the exception, which would be logged by Python to stderr.

Regardless, there is a possible hole there with unhandled exception which should be caught to prevent the listening thread from exiting.

**Comment by llunn** [comment]
@miguelgrinberg I indeed had the exception being logged to stderr.

I'm not even certain if the issue belongs to flask-socketio since it was a logic error of my own doing, but the consequence appears to be that flask-socketio hangs in some way (likely as you described) if an exception is thrown in the context of handling an event.

The snippet below should reproduce the behaviour; to do it:
1. Connect one or more clients
2. Disconnect any client
3. Send `will-not-emmit` from any connected client

```python
from flask import current_app
from flask_socketio import Namespace, emit, disconnect  # noqa: E402

class DummyNamespace(Namespace):
    def on_connect(self, auth=None):
        # UNCOMMENT TO ACTUALLY POPULATED THE CONNECTED SIDS
        #if hasattr(current_app, 'active_clients') is False:
        #    setattr(current_app, 'active_clients', {})
          
        #current_app.active_clients[request.sid] = request.sid

    def on_disconnect(self):
        # GENERATES A KEY ERROR IF CLIENT IS NOT CONNECTED; COMMENT THIS AND UNCOMMENT BELOW TO RESOLVE THE ISSUE
        current_app.active_clients.pop(request.sid)
        
        # USE THIS INSTEAD TO GUARD AGAINST KEY ERROR
        #if request.sid in current_app.active_clients:
        #    current_app.active_clients.pop(request.sid)

    def on_close(self):
        disconnect()
    
    def on_will_fail(self):
        # WILL NOT CAUSE PUB SUB AFTER KEY ERROR IS RAISED IN  on_disconnect
        emit('will-not-emmit', 'I get lost because something has hung...')
```
